src/solvers/refinement/string_refinement.cpp

```diff
@@ -249,327 +280,315 @@ static std::vector<exprt> generate_instantiations(
   for(const auto &nc_axiom : axioms.not_contains)
   {
     for(const auto &instance :
-      instantiate(nc_axiom, index_set, generator))
+      instantiate(stream, ns, nc_axiom, index_set, generator))
       lemmas.push_back(instance);
   }
   return lemmas;
 }
 
-/// List the simple expressions on which the expression depends in the
-/// `symbol_resolve` map. A simple expression is either a symbol or a
-/// constant array
-/// \param expr: an expression
-static void depends_in_symbol_map(const exprt &expr, std::vector<exprt> &accu)
+/// remove functions applications and create the necessary axioms
+/// \par parameters: an expression containing function applications
+/// \return an expression containing no function application
+static exprt substitute_function_applications(
+  exprt expr, string_constraint_generatort &generator)
 {
-  if(expr.id()==ID_if)
-  {
-    if_exprt if_expr=to_if_expr(expr);
-    depends_in_symbol_map(if_expr.true_case(), accu);
-    depends_in_symbol_map(if_expr.false_case(), accu);
-  }
-  else if(expr.id()==ID_struct)
+  for(size_t i=0; i<expr.operands().size(); ++i)
   {
-    string_exprt str=to_string_expr(expr);
-    depends_in_symbol_map(str.content(), accu);
+    // TODO: only copy when necessary
+    exprt op(expr.operands()[i]);
+    expr.operands()[i]=substitute_function_applications(op, generator);
   }
-  else
+
+  if(expr.id()==ID_function_application)
   {
-    INVARIANT(
-      expr.id()==ID_symbol || expr.id()==ID_array || expr.id()==ID_array_of,
-      "leaf in symbol resolve should be a symbol or a constant array");
-    accu.push_back(expr);
+    function_application_exprt f=to_function_application_expr(expr);
```

Comment:
`cosnt &`

Reply:
There's no difference, thanks to return value optimization.

Reply:
Maybe for clarity and future proofing then? If I added something to manipulate `f` that would benefit from `const &`, I would have to add that, but if I put it everywhere it can be I don't have to worry.